explorer/mempool.go

```diff
@@ -0,0 +1,228 @@
+		if !ok {
+			log.Infof("New Tx channel closed")
+			return
+		}
+
+		// A nil tx is the signal to stop
+		if ntx == nil {
+			return
+		}
+
+		// A tx with an empty hex is the new block signal
+		if ntx.Hex == "" {
+			exp.storeMempoolInfo()
+			exp.wsHub.HubRelay <- sigMempoolUpdate
+			continue
+		}
+
+		// Ignore this tx if it was received before the last block
+		exp.NewBlockDataMtx.Lock()
+		lastBlockTime := exp.NewBlockData.BlockTime
+		exp.NewBlockDataMtx.Unlock()
+
+		if ntx.Time < lastBlockTime {
+			continue
+		}
+
+		msgTx, err := txhelpers.MsgTxFromHex(ntx.Hex)
+		if err != nil {
+			continue
+		}
+
+		hash := msgTx.TxHash().String()
+
+		var voteInfo *VoteInfo
+		if ok := stake.IsSSGen(msgTx); ok {
+			validation, version, bits, choices, err := txhelpers.SSGenVoteChoices(msgTx, exp.ChainParams)
+			if err != nil {
+				log.Debugf("Cannot get vote choices for %s", hash)
+			} else {
+				voteInfo = &VoteInfo{
+					Validation: BlockValidation{
+						Hash:     validation.Hash.String(),
+						Height:   validation.Height,
+						Validity: validation.Validity,
+					},
+					Version: version,
+					Bits:    bits,
+					Choices: choices,
+				}
+			}
+		}
+
+		tx := MempoolTx{
+			Hash:     hash,
+			Time:     ntx.Time,
+			Size:     int32(len(ntx.Hex) / 2),
+			TotalOut: txhelpers.TotalOutFromMsgTx(msgTx).ToCoin(),
+			Type:     txhelpers.DetermineTxTypeString(msgTx),
+			VoteInfo: voteInfo,
+		}
+
+		exp.MempoolData.Lock()
+		// Add the tx to the appropriate tx slice and update the count
+		switch tx.Type {
+		case "Ticket":
+			exp.MempoolData.Tickets = append([]MempoolTx{tx}, exp.MempoolData.Tickets...)
+			exp.MempoolData.NumTickets++
+		case "Vote":
+			exp.MempoolData.Votes = append([]MempoolTx{tx}, exp.MempoolData.Votes...)
+			exp.MempoolData.NumVotes++
+		case "Regular":
+			exp.MempoolData.Transactions = append([]MempoolTx{tx}, exp.MempoolData.Transactions...)
+			exp.MempoolData.NumRegular++
+		case "Revocation":
+			exp.MempoolData.Revocations = append([]MempoolTx{tx}, exp.MempoolData.Revocations...)
+			exp.MempoolData.NumRevokes++
+		}
+
+		exp.MempoolData.LatestTransactions = append([]MempoolTx{tx}, exp.MempoolData.LatestTransactions[:len(exp.MempoolData.LatestTransactions)-1]...)
+
+		exp.MempoolData.NumAll++
+		exp.MempoolData.TotalOut += tx.TotalOut
+		exp.MempoolData.TotalSize += tx.Size
+		exp.MempoolData.FormattedTotalSize = humanize.Bytes(uint64(exp.MempoolData.TotalSize))
+
+		exp.MempoolData.Unlock()
+		exp.wsHub.HubRelay <- sigNewTx
+		exp.wsHub.NewTxChan <- &tx
+	}
+}
+
+func (exp *explorerUI) StopMempoolMonitor(txChan chan *NewMempoolTx) {
+	log.Infof("Stopping mempool monitor")
+	txChan <- nil
+}
+
+func (exp *explorerUI) StartMempoolMonitor(newTxChan chan *NewMempoolTx) {
+	go exp.mempoolMonitor(newTxChan)
+}
+
+func (exp *explorerUI) storeMempoolInfo() {
+
+	defer func(start time.Time) {
+		log.Debugf("storeMempoolInfo() completed in %v",
+			time.Since(start))
+	}(time.Now())
+
+	exp.MempoolData.Lock()
```

Comment:
I'm thinking we should move this closer to where we actually need it locked.  Even after the exp.NewBlockDataMtx.Lock() block as far as I can tell